---
title: "Feature table filtering"
author: "Yanxian Li<br><small>Department of Paraclinical Sciences<br>Faculty of Veterinary Medicine<br>Norwegian University of Life Sciences</small>"
date: "<small>`r Sys.Date()`</small>"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float: 
      collapsed: true
    code_folding: show
    theme: cerulean
    self_contained: true
  pdf_document: 
    latex_engine: xelatex
  word_document: default  
---

```{r style, echo = FALSE, message = FALSE, warning = FALSE}
require(knitr)
opts_chunk$set(message = FALSE, error = FALSE, warning = FALSE, cache = TRUE)
```

`r Hmisc::hidingTOC(buttonLabel = "Outline (hide)", tocSide = "left",  buttonSide = "left")`

# Getting ready

## Load packages

```{r}
library(here) # A Simpler Way to Find Your Files, CRAN v1.0.1
library(tidyverse) # Easily Install and Load the 'Tidyverse', CRAN v1.3.0 
library(qiime2R) # Import qiime2 artifacts to R, [github::jbisanz/qiime2R] v0.99.35  
library(phyloseq) # Analysis of high-throughput microbiome data, Bioconductor v1.34.0 
library(speedyseq) # Faster phyloseq functions, [github::mikemc/speedyseq] v0.5.3.9001
library(decontam) # Identify Contaminants in Sequencing Data, Bioconductor v1.10.0
library(cowplot) # Streamlined Plot Theme and Plot Annotations for 'ggplot2', CRAN v1.1.1  
library(ggstatsplot) # 'ggplot2' Based Plots with Statistical Details, CRAN v0.6.8
library(RColorBrewer) # ColorBrewer Palettes, CRAN v1.1-2 
library(biomformat) # An interface package for the BIOM file format, Bioconductor v1.18.0
```

## Load functions

```{r}
source(here("code/functions/plot_frequency.R"))
source(here("code/functions/plot_prevalence.R"))
```

## Data wrangling

### Metadata

```{r}
mtd <- read_tsv(here("data/metadata.tsv"), comment = "#q2") 
mtd <- rename(mtd, SampleID = "#SampleID") %>%
  # create new variables for identifying contaminating features
  mutate(IsNeg = ifelse(grepl("blank", SampleType), TRUE, FALSE),
         DilutedDNAConcentration = ifelse(
           SampleType == "PCR-blank", 
           mean(filter(mtd, SampleType == "Extraction-blank")$DilutedDNAConcentration),
           DilutedDNAConcentration),
         DNAConcentration = ifelse(
           SampleType == "PCR-blank", 
           DilutedDNAConcentration, 
           DilutedDNAConcentration * DNADilutionFactor),
         DNAExtractionBatch = as.character(DNAExtractionBatch),
         Diet = factor(Diet, levels = c("REF", "IM")),
         Compartment = factor(Compartment, levels = c("PI", "DI")),
         SampleType = factor(SampleType, levels = c("REF-PID", "REF-PIM", "REF-DID", "REF-DIM", 
                                                    "IM-PID", "IM-PIM", "IM-DID", "IM-DIM", 
                                                    "Feed", "Water", "Extraction-blank", 
                                                    "PCR-blank", "Mock"))) 
```

### Feature table

```{r}
table <- read_qza(here("data/intermediate/qiime2/asv/table.qza"))
count_tab <- table$data %>% as.data.frame() 
```

### Taxonomy

```{r}
taxonomy <- read_qza(here("data/intermediate/qiime2/asv/taxonomy-silva132.qza"))
tax_tab <- taxonomy$data %>% 
  as.data.frame() %>%
  mutate(Taxon = gsub("D_0", "k", Taxon), Taxon = gsub("D_1", "p", Taxon),
         Taxon = gsub("D_2", "c", Taxon), Taxon = gsub("D_3", "o", Taxon),
         Taxon = gsub("D_4", "f", Taxon), Taxon = gsub("D_5", "g", Taxon),
         Taxon = gsub("D_6", "s", Taxon)) %>%
  separate(Taxon, sep = ";", c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>% 
  column_to_rownames("Feature.ID") %>%
  select(-Confidence)
```

### Phyloseq object

```{r}
ps <- phyloseq(sample_data(column_to_rownames(mtd, "SampleID")),
               otu_table(as.matrix(count_tab), taxa_are_rows = TRUE),
               tax_table(as.matrix(tax_tab)))

# total sum scaling
ps_tss <- transform_sample_counts(ps, function(x){
  if(sum(x) == 0){ # avoid producing naN when sample count is zero
    x 
  } else {
    x / sum(x)
  }
}
) 
```

# Taxonomy-based filtering

Remove features without a phylum-level annotation and those assigned as chloroplast or mitochondria. Note that the taxonomic labels are database specific and may change in different versions of the same database. Make sure you're using the correct taxonomic labels to remove chloroplast and mitochondria.

```{r}
ps_tss <- subset_taxa(ps_tss, !is.na(Phylum) & !Phylum %in% c("", "uncharacterized")) %>%
  subset_taxa(Order != "o__Chloroplast"|is.na(Order)) %>%
  subset_taxa(Family != "f__Mitochondria"|is.na(Family))
```

# Prevalence-based filtering

Features that show up in only one or a few samples may not represent real biological diversity but rather PCR/sequencing errors (such as PCR chimeras) or reagent contaminants.

```{r}
ps_tss <- subset_samples(ps_tss, !SampleType %in% c("Extraction-blank", "PCR-blank")) %>%
  # remove features present in only one sample
  filter_taxa(., function(x) sum(x > 0) > 1, TRUE) %>%
  taxa_names() %>%
  prune_taxa(ps_tss)
```

# Filter contaminants

Reagent contamination is unavoidable and study-specific, which can critically influence sequencing-based microbiome analyses. The screening of reagent contaminants will be based on two typical characteristics of contaminating sequences as outlined in the paper [Simple statistical identification and removal of contaminant sequences in marker-gene and metagenomics data](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0605-2): they are likely to have frequencies that inversely correlate with sample DNA concentration and are likely to have higher prevalence in control samples than in true samples. The authors developed an R package, [*decontam*](https://github.com/benjjneb/decontam), for removing contaminating sequences in the marker-gene and shotgun metagenomics data. The package, however, does not make use of positive controls for the identification of contaminating sequences. As removing of features may critically affect downstream analyses, we'll do it by manual screening based on the aforementioned principles.

## Inspect bacterial DNA concentration

Before we proceed with the identification of contaminating features, let's check the bacterial DNA concentration and see if there're systemic differences among different types of samples. In this study, the bacterial DNA was quantified by qPCR using [a universal primer set](https://www.nature.com/articles/nature24460) targeting the bacterial 16S rRNA gene. The standard curve was generated using the [Femto bacterial DNA standards](https://www.zymoresearch.com/collections/femto-dna-quantification-kits/products/femto-bacterial-dna-quantification-kit) purified from the *E. coli* strain JM109.

First of all, let's look at the Cq values of the DNA templates used for the amplicon PCR. Note that the Cq values only reflect the bacterial DNA concentration in the DNA templates used for the amplicon PCR. The extracted genomic DNA of some samples was diluted before running the amplicon PCR.

```{r}
filter(mtd, SampleType!= "PCR-blank") %>%
  ggplot(aes(x = SampleType, y = DilutedDNACqValue)) +
    geom_jitter(aes(colour = SampleType), shape = 16, position = position_jitter(0.1)) +
    # add mean plus/minus 1SD 
    stat_summary(fun.data = "mean_sdl", fun.args = list(mult = 1), geom = "pointrange") + 
    coord_flip() +
    scale_x_discrete(limits = rev(levels(mtd$SampleType))) +
    labs(x = "Sample type", y = "Cq value", 
         title = "Cq vlaues of DNA templates used for the amplicon PCR") +
    theme_minimal(base_size = 12) +
    theme(legend.position = "none")
```

Now, let's look at the original bacterial DNA concentration of the extracted genomic DNA. Note that the bacterial DNA concentration was not normalized. Thus, these numbers do not reflect the bacterial DNA concentration of the samples but rather the extracted genomic DNA.

```{r}
filter(mtd, SampleType!= "PCR-blank") %>%
  ggplot(aes(x = SampleType, y = log10(DNAConcentration * 1000000))) +
    geom_jitter(aes(colour = SampleType), shape = 16, position = position_jitter(0.1)) +
    stat_summary(fun.data = "mean_sdl", fun.args = list(mult = 1), geom = "pointrange") + 
    coord_flip() +
    scale_x_discrete(limits = rev(levels(mtd$SampleType))) +
    labs(x = "Sample type", y = "log10(DNA concentration) (fg/Î¼l)", 
         title = "Bacterial DNA concentration in the extracted genomic DNA") +
    theme_minimal(base_size = 12) +
    theme(legend.position = "none")
```

## Identify PCR contaminants

Based on the taxa barplot generated by the QIIME2, the majority of contaminating features in the negative controls have originated from the amplicon PCR. Thus, we'll start with filtering contaminants introduced during the amplicon PCR.

### Batch1

#### Data wrangling

Subset samples and features for the screening of contaminating features.

```{r}
# subset samples from amplicon PCR batch1 
ps_tss_pcr1 <- subset_samples(ps_tss, PCRBatch == 1)

# make a dataframe containing features present in the negative controls of PCR 
decontam_pcr1 <- ps_tss_pcr1 %>%
  # the following 4 lines remove features not present in the negative controls of PCR
  subset_samples(SampleType == "PCR-blank") %>%
  filter_taxa(., function(x) sum(x > 0) > 0, TRUE) %>%
  taxa_names() %>%
  prune_taxa(ps_tss_pcr1) %>%
  # convert the phyloseq object into a tidy style dataframe
  psmelt() %>%
  unite("tax", Kingdom:Species, sep = "; ") %>%
  arrange(OTU, SampleType, SampleName) %>%
  mutate(Abundance = Abundance * 100,
         SampleName = factor(SampleName, unique(SampleName))) 

# sort the features based on their abundance in the negative controls of PCR  
otu_sorted.pcr1 <- filter(decontam_pcr1, SampleType == "PCR-blank") %>%
  group_nest(OTU) %>%
  mutate(taxa_sum = map(data, ~sum(.x$Abundance))) %>%
  unnest(taxa_sum) %>%
  arrange(desc(taxa_sum))

# arrange features by their abundance in the negative controls of PCR
decontam_pcr1 <- mutate(decontam_pcr1, OTU = factor(OTU, otu_sorted.pcr1$OTU)) %>% arrange(OTU)
```

#### Prevalence-based classification

Here we use barplots to visualize the abundance and prevalence of the features found in the negative controls of PCR.

```{r, results='hide'}
# split the dataframe by feature ID
decontam_pcr1_spl1 <- group_split(decontam_pcr1, OTU)

# make barplots
pdf(here("data/intermediate/filtering/decontam/prevalence_pcr1.pdf"), width = 16, height = 10) 

lapply(seq_along(decontam_pcr1_spl1), function(x){
  # make a bar plot without mock
  p1 <- filter(decontam_pcr1_spl1[[x]], SampleType != "Mock") %>%
    plot_prevalence(x = SampleName, y = Abundance, bar_color = SampleType,
                    xlab = "Sample name", ylab = "Relative abundance (%)", 
                    title = unique(decontam_pcr1_spl1[[x]][, "tax"]))
  
  # make a bar plot with mock only
  p2 <- filter(decontam_pcr1_spl1[[x]], SampleType == "Mock") %>%
    plot_prevalence(x = SampleName, y = Abundance, bar_color = SampleType, xlab = "", ylab = "")
  
  # assemble plots
  plot_grid(p1, p2 + theme(legend.position = "none"), nrow = 1, align = 'h', axis = "bt", rel_widths = c(13, 1))
  }
)

dev.off() 
```

#### Frequency-based classification

Here we visualize the correlation between the baterial DNA concentration and the abundance of features found in the negative controls of PCR. For convenience, we'll use the Cq values as the proxy of baterial DNA concentration. Features showing positive correlations with Cq values are potential contaminating features.

```{r, results='hide'}
# exclude samples with zero count
decontam_pcr1_spl2 <- decontam_pcr1 %>% 
  filter(Abundance != 0 & SampleType != "PCR-blank") %>% 
  group_split(OTU)

# make correlation plots
pdf(here("data/intermediate/filtering/decontam/frequency_pcr1.pdf"), width = 16, height = 12) 

lapply(seq_along(decontam_pcr1_spl2), function(x){
  # extract taxonomy as the plot title
  main <- ggdraw() + draw_label(unique(decontam_pcr1_spl2[[x]]$tax), fontface='bold')
  
  # correlation with all samples
  p1 <- plot_frequency(decontam_pcr1_spl2[[x]], x = DilutedDNACqValue, y = Abundance, 
                       type = "nonparametric", point_color = SampleType, xlab = "Cq value", 
                       ylab = "Relative abundance (%)", title = "Correlation with all samples")
 
   # correlation within each sample type
  p2 <- plot_frequency(decontam_pcr1_spl2[[x]], x = DilutedDNACqValue, y = Abundance, 
                       facet_wrap = SampleType, facet_ncol = 2, type = "nonparametric", 
                       xlab = "Cq value", ylab = "Relative abundance (%)")
  
  # assemble plots
  plot_grid(main, p1, p2, ncol = 1, rel_heights = c(1, 10, 12))
  }
)

dev.off() 
```

#### Gather contaminating features

After inspecting the feature prevalence barplots and the correlation plots, the following features are considered as contaminants:

```{r}
contam_pcr1 <- select(decontam_pcr1, OTU, tax) %>%
  distinct() %>%
  # remove Pseudomonas aeruginosa so that it won't be selected as a contaminating feature
  slice(-74) %>% 
  # gather contaminating features using their taxonomic labels
  filter(grepl("Pseudomonas|Halomonas|Shewanella|Burkholderiaceae|Candidatus|Pelobacter|
               |Paracoccus|Cutibacterium|Acinetobacter|Cloacibacterium", tax))
```

Check the distibution of contaminating features among samples.

```{r, fig.width=16, fig.height=10}
prune_taxa(taxa_names(ps_tss) %in% contam_pcr1$OTU, ps_tss) %>%
  plot_bar(x = "SampleName", fill = "Genus", title = "Distibution of contaminating features: PCR batch1") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +  
  scale_fill_brewer(palette = "Paired") +
  theme_bw() +
  theme(legend.position = "bottom", axis.text.x = element_text(angle = 90, hjust = 1))
```

### Batch2

#### Data wrangling

```{r}
# subset samples from amplicon PCR batch2 
ps_tss_pcr2 <- subset_samples(ps_tss, PCRBatch == 2)

# make a dataframe containing features present in the negative controls of PCR 
decontam_pcr2 <- ps_tss_pcr2 %>%
  subset_samples(SampleType == "PCR-blank") %>%
  filter_taxa(., function(x) sum(x > 0) > 0, TRUE) %>%
  taxa_names() %>%
  prune_taxa(ps_tss_pcr2) %>%
  psmelt() %>%
  unite("tax", Kingdom:Species, sep = "; ") %>%
  arrange(OTU, SampleType, SampleName) %>%
  mutate(Abundance = Abundance * 100,
         SampleName = factor(SampleName, unique(SampleName))) 

# sort the features based on their abundance in the negative controls of PCR  
otu_sorted.pcr2 <- filter(decontam_pcr2, SampleType == "PCR-blank") %>%
  group_nest(OTU) %>%
  mutate(taxa_sum = map(data, ~sum(.x$Abundance))) %>%
  unnest(taxa_sum) %>%
  arrange(desc(taxa_sum))

# arrange features by their abundance in the negative controls of PCR
decontam_pcr2 <- mutate(decontam_pcr2, OTU = factor(OTU, otu_sorted.pcr2$OTU)) %>% arrange(OTU)
```

#### Prevalence-based classification

```{r, results='hide'}
decontam_pcr2_spl1 <- group_split(decontam_pcr2, OTU)

pdf(here("data/intermediate/filtering/decontam/prevalence_pcr2.pdf"), width = 16, height = 10) 

lapply(seq_along(decontam_pcr2_spl1), function(x){
  p1 <- filter(decontam_pcr2_spl1[[x]], SampleType != "Mock") %>%
    plot_prevalence(x = SampleName, y = Abundance, bar_color = SampleType,
                    xlab = "Sample name", ylab = "Relative abundance (%)", 
                    title = unique(decontam_pcr2_spl1[[x]][, "tax"]))
  
  p2 <- filter(decontam_pcr2_spl1[[x]], SampleType == "Mock") %>%
    plot_prevalence(x = SampleName, y = Abundance, bar_color = SampleType, xlab = "", ylab = "")
  
  plot_grid(p1, p2 + theme(legend.position = "none"), nrow = 1, align = 'h', axis = "bt", rel_widths = c(13, 1))
  }
)

dev.off()
```

#### Frequency-based classification

```{r, results='hide'}
decontam_pcr2_spl2 <- decontam_pcr2 %>% 
  filter(Abundance != 0 & SampleType != "PCR-blank") %>% 
  group_split(OTU)

pdf(here("data/intermediate/filtering/decontam/frequency_pcr2.pdf"), width = 16, height = 12) 

lapply(seq_along(decontam_pcr2_spl2), function(x){
  main <- ggdraw() + draw_label(unique(decontam_pcr2_spl2[[x]]$tax), fontface='bold')
  
  p1 <- plot_frequency(decontam_pcr2_spl2[[x]], x = DilutedDNACqValue, y = Abundance, 
                       type = "nonparametric", point_color = SampleType, xlab = "Cq value", 
                       ylab = "Relative abundance (%)", title = "Correlation with all samples")

  p2 <- decontam_pcr2_spl2[[x]] %>%
    filter(SampleType %in% c("REF-PIM", "REF-DIM", "IM-PIM", "IM-DIM", "Water", "Feed")) %>%
    plot_frequency(x = DilutedDNACqValue, y = Abundance, facet_wrap = SampleType, facet_ncol = 2, 
                   type = "nonparametric", xlab = "Cq value", ylab = "Relative abundance (%)")
  
  plot_grid(main, p1, p2, ncol = 1, rel_heights = c(1, 10, 15))
  }
)

dev.off() 
```

#### Gather contaminating features

The majority of reads in two of the three negative controls (PCR) were filtered during the sequence denoising process by the DADA2 algorithm (3 sequences left in LB2a and 1 sequence in LB2c). After inspecting the sequence prevalence barplots, the remaining negative control, LB2b, is likely cross-contaminated by the sample IMPm206. Only one feature belonging to the *Bradyrhizobium* genus is considered as a reagent contaminant.

```{r}
contam_pcr2 <- select(decontam_pcr2, OTU, tax) %>%
  distinct() %>%
  filter(grepl("Bradyrhizobium", tax))
```

Check the distibution of contaminating features among samples

```{r, fig.width=16, fig.height=10}
prune_taxa(taxa_names(ps_tss) %in% contam_pcr2$OTU, ps_tss) %>%
  plot_bar(x = "SampleName", fill = "Genus", title = "Distibution of contaminating features: PCR batch2") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  scale_fill_brewer(palette = "Paired") +
  theme_bw() +
  theme(legend.position = "bottom", axis.text.x = element_text(angle = 90, hjust = 1))
```

### Batch3

#### Data wrangling

The amplification of some REF-PIM/REF-DIM samples failed during the 2nd batch of amplicon PCR. Thus, these samples were amplified again in the PCR batch3. To make it easier for the identification of contaminating features, we'll also include the REF-PIM/REF-DIM samples sucessfully amplified in the PCR batch2 when making the feature prevalence barplots.

```{r}
# subset samples 
ps_tss_pcr3 <- subset_samples(ps_tss, SampleType %in% c("REF-PIM", "REF-DIM", "PCR-blank")) %>%
  subset_samples(!SampleName %in% c("LB1a", "LB1b", "LB1c", "LB2a", "LB2b", "LB2c"))

# make a dataframe containing features present in the negative controls of PCR 
decontam_pcr3 <- ps_tss_pcr3 %>%
  subset_samples(SampleType == "PCR-blank") %>%
  filter_taxa(., function(x) sum(x > 0) > 0, TRUE) %>%
  taxa_names() %>%
  prune_taxa(ps_tss_pcr3) %>%
  psmelt() %>%
  unite("tax", Kingdom:Species, sep = "; ") %>%
  arrange(OTU, SampleType, PCRBatch, SampleName) %>%
  mutate(Abundance = Abundance * 100,
         SampleName = factor(SampleName, unique(SampleName)),
         PCRBatch = factor(PCRBatch, c(2,3))) 

# sort the features based on their abundance in the negative controls of PCR  
otu_sorted.pcr3 <- filter(decontam_pcr3, SampleType == "PCR-blank") %>%
  group_nest(OTU) %>%
  mutate(taxa_sum = map(data, ~sum(.x$Abundance))) %>%
  unnest(taxa_sum) %>%
  arrange(desc(taxa_sum))

# arrange features by their abundance in the negative controls of PCR
decontam_pcr3 <- mutate(decontam_pcr3, OTU = factor(OTU, otu_sorted.pcr3$OTU)) %>% arrange(OTU)
```

#### Prevalence-based classification

```{r, results='hide'}
decontam_pcr3_spl1 <- group_split(decontam_pcr3, OTU)

pdf(here("data/intermediate/filtering/decontam/prevalence_pcr3.pdf"), width = 16, height = 10) 

lapply(seq_along(decontam_pcr3_spl1), function(x){
  plot_prevalence(decontam_pcr3_spl1[[x]], x = SampleName, y = Abundance, bar_color = PCRBatch,
                  facet_wrap = SampleType, scale_facet = "free_x", xlab = "Sample name", 
                  ylab = "Relative abundance (%)", title = unique(decontam_pcr3_spl1[[x]][, "tax"]))
 } 
)

dev.off() 
```

#### Frequency-based classification

```{r, results='hide'}
decontam_pcr3_spl2 <- decontam_pcr3 %>% 
  filter(PCRBatch == 3 & Abundance != 0) %>% 
  group_split(OTU)

pdf(here("data/intermediate/filtering/decontam/frequency_pcr3.pdf"), width = 16, height = 12) 

lapply(seq_along(decontam_pcr3_spl2), function(x){
  main <- ggdraw() + draw_label(unique(decontam_pcr3_spl2[[x]]$tax), fontface='bold')
  
  p1 <- plot_frequency(decontam_pcr3_spl2[[x]], x = DilutedDNACqValue, y = Abundance, 
                       type = "nonparametric", point_color = SampleType, xlab = "Cq value", 
                       ylab = "Relative abundance (%)", title = "Correlation with all samples", )

  p2 <- plot_frequency(decontam_pcr3_spl2[[x]], x = DilutedDNACqValue, y = Abundance, 
                       facet_wrap = SampleType, facet_ncol = 2, type = "nonparametric", 
                       xlab = "Cq value", ylab = "Relative abundance (%)")
  
  plot_grid(main, p1, p2, ncol = 1, rel_heights = c(1, 10, 10))
  }
)

dev.off() 
```

#### Gather contaminating features

While some common reagent contaminants are present in the negative controls of PCR, the correlaion plots didn't show clear associations between the Cq values and the abundance of contaminating features. This's possibly due to the lower bacterial biomass in the mucosal samples that cann't be accurately quantified by qPCR. The screening of contaminating features for the PCR batch3 is mostly based on the prior knowledge of common reagent contaminants found in our lab and in the literature ( recently reviewed by [Eisenhofer et al., 2019](https://www.sciencedirect.com/science/article/pii/S0966842X18302531)). The following features are considered as contaminants:

```{r}
contam_pcr3 <- select(decontam_pcr3, OTU, tax) %>%
  distinct() %>%
  filter(grepl("Burkholderiaceae|Xanthobacteraceae|Chitinophagaceae|
               |Caulobacteraceae|Sphingomonas|Cutibacterium", tax))
```

Check the distibution of contaminating features among samples.

```{r, fig.width=16, fig.height=10}
prune_taxa(taxa_names(ps_tss) %in% contam_pcr3$OTU, ps_tss) %>%
  plot_bar(x = "SampleName", fill = "Genus", title = "Distibution of contaminating features: PCR batch3") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  theme_bw() +
  theme(legend.position = "bottom", axis.text.x = element_text(angle = 90, hjust = 1))
```

## Identify DNA extraction contaminants

### Data wrangling

```{r}
# gather the PCR contaminants
contam_pcr <- bind_rows(contam_pcr1, contam_pcr2, contam_pcr3)

# remove the PCR contaminants and negative controls
ps_tss_no_pcr.contam_lb <- prune_taxa(!taxa_names(ps_tss) %in% contam_pcr$OTU, ps_tss) %>%
  subset_samples(SampleType != "PCR-blank")

# remove features not present in the negative controls of DNA extraction 
ps_tss_decontam_dna <- subset_samples(ps_tss_no_pcr.contam_lb, SampleType == "Extraction-blank") %>%
  filter_taxa(., function(x) sum(x > 0) > 0, TRUE) %>%
  taxa_names() %>%
  prune_taxa(ps_tss_no_pcr.contam_lb) 
```

### Initial screening by *decontam*

We'll use the *decontam* package to screen potential contaminating features introduced during the DNA extraction. Since this's an initial screening process, we set the `method = "either"`(contaminants are called if identified by either the frequency or prevalence methods) and `threshold = 0.5`. Setting the threshold at 0.5 has a particularly simple interpretation: in the frequency approach, sequence features would be classified as contaminants if the contaminant model is a better fit than the non-contaminant model, and in the prevalence approach, sequence features would be classified as contaminants if present in a higher fraction of negative controls than true samples. Thus, these settings result in much more aggressive classifications compared to the default method. Also note that we pool all the samples to identify the potential contaminants rather than running the algorithm on a per batch basis. This's because each batch of DNA extraction only contains 8 samples plus a negative control, which actually resulted in worse classifications than pooling samples.

```{r}
# identify potential contaminating features
contam_dna.init <- isContaminant(ps_tss_decontam_dna, conc = "DNAConcentration", neg = "IsNeg", 
                                 method = "either", threshold = 0.5, normalize = FALSE)

# check the number of potential contaminating features
table(contam_dna.init$contaminant)
```

### Further examination

Data wrangling.

```{r}
decontam_dna <- prune_taxa(contam_dna.init$contaminant, ps_tss_decontam_dna) %>%
  psmelt() %>%
  unite("tax", Kingdom:Species, sep = "; ") %>%
  arrange(OTU, SampleType, SampleName) %>%
  mutate(Abundance = Abundance * 100,
         SampleName = factor(SampleName, unique(SampleName))) 

otu_sorted.dna <- filter(decontam_dna, SampleType == "Extraction-blank") %>%
  group_nest(OTU) %>%
  mutate(taxa_sum = map(data, ~sum(.x$Abundance))) %>%
  unnest(taxa_sum) %>%
  arrange(desc(taxa_sum))

decontam_dna <- mutate(decontam_dna, OTU = factor(OTU, otu_sorted.dna$OTU)) %>% arrange(OTU)
```

#### Prevalence-based classification

```{r, results='hide'}
decontam_dna_spl1 <- group_split(decontam_dna, OTU)

pdf(here("data/intermediate/filtering/decontam/prevalence_dna.pdf"), width = 16, height = 10) 

lapply(seq_along(decontam_dna_spl1), function(x){
  p1 <- filter(decontam_dna_spl1[[x]], SampleType != "Mock") %>%
    plot_prevalence(x = SampleName, y = Abundance, bar_color = SampleType,
                    xlab = "Sample name", ylab = "Relative abundance (%)", 
                    title = unique(decontam_dna_spl1[[x]][, "tax"])) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
  
  p2 <- filter(decontam_dna_spl1[[x]], SampleType == "Mock") %>%
    plot_prevalence(x = SampleName, y = Abundance, bar_color = SampleType, xlab = "", ylab = "") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
  
  plot_grid(p1, p2 + theme(legend.position = "none"), nrow = 1, align = 'h', axis = "bt", rel_widths = c(13, 1))
  }
)

dev.off() 
```

#### Frequency-based classification

```{r, results='hide'}
decontam_dna_spl2 <- filter(decontam_dna, Abundance != 0) %>% group_split(OTU)

pdf(here("data/intermediate/filtering/decontam/frequency_dna.pdf"), width = 16, height = 10) 

lapply(seq_along(decontam_dna_spl2), function(x){
  plot_frequency(data = decontam_dna_spl2[[x]], x = DilutedDNACqValue, y = Abundance, 
                 point_color = SampleType, xlab = "Cq value", ylab = "Relative abundance (%)", 
                 title = unique(decontam_dna_spl2[[x]]$tax), type = "nonparametric")
  }
)

dev.off() 
```

#### Gather contaminating features

The following features are considered as contaminants:

```{r}
contam_dna <- select(decontam_dna, OTU, tax) %>%
  distinct() %>%
  # remove Pseudomonas aeruginosa so that it won't be selected as a contaminating feature
  slice(-47) %>% 
  filter(grepl("Pseudomonas|Halomonas|Shewanella|Burkholderiaceae|Bacillus okhensis|
               |Rhodobacteraceae|Acinetobacter|Microbacterium|Cloacibacterium|Vogesella|
               |Marinobacter|Methylobacterium|Dickeya", tax))
```

Check the distibution of contaminating features among samples.

```{r, fig.width=16, fig.height=10}
prune_taxa(taxa_names(ps_tss) %in% contam_dna$OTU, ps_tss) %>%
  plot_bar(x = "SampleName", fill = "Genus", title = "Distibution of contaminating features: DNA extraction") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  theme_bw() +
  theme(legend.position = "bottom", axis.text.x = element_text(angle = 90, hjust = 1))
```

Instead of a uniform distribution among all samples, we can see that the purported DNA extraction contaminants are only present in samples from the first batch of amplicon PCR. This indicates that these remaining contaminants in the DNA extraction blanks are from the amplicon PCR which were absent in the negative controls of the amplicon PCR. In summary, the contaminants introduced during the amplicon PCR dwarfs those from the DNA extraction.

## Identify cross-contamination in the mock

Besides reagent and laboratory contamination, between-sample contamination, whether due to the "barcode swapping" and/or physical cross-contamination of samples, is also known to occur at low levels in molecular-based microbiome studies. The mock, [ZymoBIOMICS D6300](https://www.zymoresearch.com/collections/zymobiomics-microbial-community-standards/products/zymobiomics-microbial-community-standard), comes with a guaranteed impurity level of \< 0.01% (by DNA abundance). As long as we observe any alien taxa present at \> 0.01% in the mock, we can conclude that they are introduced during by the workflow. Thus, identifying contaminants in the mock samples is more straightforward, which can be done based on the taxonomic labels of the features and their abundance across all the samples.

### Data wrangling

```{r}
# remove reagent contaminants identified in the previous steps
contam <- bind_rows(contam_pcr, contam_dna)
ps_tss_nocontam <- prune_taxa(!taxa_names(ps_tss) %in% contam$OTU, ps_tss)
  
# remove features not present in the mock samples
decontam_mock <- subset_samples(ps_tss_nocontam, SampleType == "Mock") %>%
  filter_taxa(., function(x) sum(x > 0) > 0, TRUE) %>%
  taxa_names() %>%
  prune_taxa(ps_tss_nocontam) %>%
  psmelt() %>%
  unite("tax", Kingdom:Species, sep = "; ") %>%
  arrange(OTU, SampleType, SampleName) %>%
  mutate(Abundance = Abundance * 100,
         SampleName = factor(SampleName, unique(SampleName))) 

otu_sorted.mock <- filter(decontam_mock, SampleType == "Mock") %>%
  group_nest(OTU) %>%
  mutate(taxa_sum = map(data, ~sum(.x$Abundance))) %>%
  unnest(taxa_sum) %>%
  arrange(desc(taxa_sum))

decontam_mock <- mutate(decontam_mock, OTU = factor(OTU, otu_sorted.mock$OTU)) %>% arrange(OTU)
```

### Prevalence-based classification

```{r, results='hide'}
decontam_mock_spl <- group_split(decontam_mock, OTU)

pdf(here("data/intermediate/filtering/decontam/prevalence_mock.pdf"), width = 16, height = 10) 

lapply(seq_along(decontam_mock_spl), function(x){
  plot_prevalence(data = decontam_mock_spl[[x]], x = SampleName, y = Abundance, 
                  bar_color = SampleType, xlab = "Sample name", ylab = "Relative abundance (%)", 
                  title = unique(decontam_mock_spl[[x]][, "tax"])) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
 }
)

dev.off() 
```

### Gather contaminating features

In addition to contaminants removed in the previous steps, the following features are considered as contaminants in the mock samples as well:

```{r}
cross_contam_mock <- select(decontam_mock, OTU, tax) %>%
  distinct() %>%
  slice(60:nrow(.))
```

Check the distibution of contaminating features.

```{r, fig.width=16, fig.height=10}
prune_taxa(taxa_names(ps_tss) %in% cross_contam_mock$OTU, ps_tss) %>%
  plot_bar(x = "SampleName", fill = "Genus", title = "Cross-contaminating features in the mock samples") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
  theme_bw() +
  theme(legend.position = "bottom", axis.text.x = element_text(angle = 90, hjust = 1))
```

## Remove contaminants

```{r}
# remove reagent contaminants from all samples.
ps_nocontam <- prune_taxa(taxa_names(ps_tss_nocontam), ps) 

# remove between-sample contaminants from the mock samples
ps_nocontam_mock <- subset_samples(ps_nocontam, SampleType == "Mock")
ps_nocontam_mock <- prune_taxa(!taxa_names(ps_nocontam_mock) %in% cross_contam_mock$OTU, ps_nocontam_mock)

# merge phyoloseq object
ps_nocontam <- subset_samples(ps_nocontam, SampleType != "Mock") %>%
  merge_phyloseq(ps_nocontam_mock)

# remove features with zero count in all samples
ps_nocontam <- filter_taxa(ps_nocontam, function(x) sum(x > 0) > 0, TRUE)
```

# Export data

Export contaminating features as Table S1.

```{r}
contam_neg <- prune_taxa(taxa_names(ps_tss) %in% contam$OTU, ps_tss) %>%
  subset_samples(SampleType %in% c("Extraction-blank", "PCR-blank")) %>%
  psmelt() %>%
  rename(FeatureID = OTU) %>%
  unite("Taxonomy", Kingdom:Species, sep = "; ") %>%
  select(FeatureID, Taxonomy, Abundance, SampleName) %>%
  mutate(FeatureID = factor(FeatureID, levels = unique(FeatureID))) %>%
  arrange(FeatureID, SampleName) %>%
  pivot_wider(names_from = SampleName, values_from = Abundance) 

# export data
write.csv(contam_neg, here("result/table/Table S1.csv"), row.names = FALSE)
```

Export filtered feature table.

```{r export_contam}
otu_table(ps_nocontam) %>% 
  as("matrix") %>%
  make_biom() %>%
  write_biom(here("data/intermediate/filtering/table-filtered.biom"))
```

# Session information

```{r session-info}
sessionInfo()
```
